src/local/room_coordinate.rs

```diff
@@ -0,0 +1,136 @@
+use std::convert::TryFrom;
+use std::fmt;
+
+use serde::{Deserialize, Deserializer, Serialize, Serializer, de};
+
+pub const ROOM_SIZE: u8 = 50;
+pub const ROOM_AREA: usize = (ROOM_SIZE as usize) * (ROOM_SIZE as usize);
+
+#[derive(Debug, Clone, Copy)]
+pub struct OutOfBoundsError(u8);
+
+impl fmt::Display for OutOfBoundsError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Out of bounds coordinate: {}", self.0)
+    }
+}
+
+#[inline]
+pub fn xy_to_linear_index(xy: RoomXY) -> usize {
+    ((xy.x.0 as usize) * (ROOM_SIZE as usize)) + (xy.y.0 as usize)
+}
+
+#[inline]
+pub fn linear_index_to_xy(idx: usize) -> RoomXY {
+    assert!(idx < ROOM_AREA, "Out of bounds index: {}", idx);
+    // SAFETY: bounds checking above ensures both are within range.
+    RoomXY {
+        x: unsafe { RoomCoordinate::unchecked_new((idx / (ROOM_SIZE as usize)) as u8) },
+        y: unsafe { RoomCoordinate::unchecked_new((idx % (ROOM_SIZE as usize)) as u8) }
+    }
+}
+
+#[derive(Debug, Hash, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct RoomCoordinate(u8);
```

Comment:
Likely want: ``#[serde(transparent)]``

Reply:
Don't we want to do the validation on deserializing too? That's why it has the custom impls for serialize and deserialize.

Reply:
Good point - I think you get the same result given you just serialize the internal value.